### PR TITLE
fix:full outer join with string_agg function panic

### LINKED ISSUE #12913: bug: full outer join with `string_agg` function panic
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

v1.2.118-nightly

### What's Wrong?

full outer join with `string_agg` function panic

```
panicked at 'assertion failed: `(left == right)`
  left: `2`,
 right: `5`', /home/b41sh/.cargo/git/checkouts/arrow2-8a2ad61d97265680/f609d0c/src/bitmap/bitmap_ops.rs:170:9
   0: common_tracing::panic_hook::log_panic
             at ./src/common/tracing/src/panic_hook.rs:35:21
   1: common_tracing::panic_hook::set_panic_hook::{{closure}}
             at ./src/common/tracing/src/panic_hook.rs:30:9
   2: call<(&core::panic::panic_info::PanicInfo), (dyn core::ops::function::Fn<(&core::panic::panic_info::PanicInfo), Output=()> + core::marker::Send + core::marker::Sync), alloc::alloc::Global>
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/alloc/src/boxed.rs:1999:9
   3: rust_panic_with_hook
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:709:13
   4: {closure#0}
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:597:13
   5: __rust_end_short_backtrace<std::panicking::begin_panic_handler::{closure_env#0}, !>
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/sys_common/backtrace.rs:151:18
   6: begin_panic_handler
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:593:5
   7: panic_fmt
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/panicking.rs:67:14
   8: assert_failed_inner
   9: core::panicking::assert_failed
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/panicking.rs:228:5
  10: arrow2::bitmap::bitmap_ops::and
             at /home/b41sh/.cargo/git/checkouts/arrow2-8a2ad61d97265680/f609d0c/src/bitmap/bitmap_ops.rs:170:9
  11: common_expression::function::passthrough_nullable::{{closure}}
             at ./src/query/expression/src/function.rs:656:29
  12: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/alloc/src/boxed.rs:1999:9
  13: common_expression::evaluator::Evaluator::partial_run
             at ./src/query/expression/src/evaluator.rs:164:30
  14: common_expression::evaluator::Evaluator::run
             at ./src/query/expression/src/evaluator.rs:93:9
  15: common_sql::evaluator::block_operator::BlockOperator::execute
             at ./src/query/sql/src/evaluator/block_operator.rs:96:38
  16: <common_sql::evaluator::block_operator::CompoundBlockOperator as common_pipeline_transforms::processors::transforms::transform::Transform>::transform::{{closure}}
             at ./src/query/sql/src/evaluator/block_operator.rs:511:47
  17: core::iter::traits::iterator::Iterator::try_fold
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/iter/traits/iterator.rs:2303:21
  18: <common_sql::evaluator::block_operator::CompoundBlockOperator as common_pipeline_transforms::processors::transforms::transform::Transform>::transform
             at ./src/query/sql/src/evaluator/block_operator.rs:509:9
  19: <common_pipeline_transforms::processors::transforms::transform::Transformer<T> as common_pipeline_core::processors::processor::Processor>::process
             at ./src/query/pipeline/transforms/src/processors/transforms/transform.rs:111:30
  20: <alloc::boxed::Box<T> as common_pipeline_core::processors::processor::Processor>::process
             at ./src/query/pipeline/core/src/processors/processor.rs:159:9
  21: common_pipeline_core::processors::processor::ProcessorPtr::process
             at ./src/query/pipeline/core/src/processors/processor.rs:127:9
  22: databend_query::pipelines::executor::executor_worker_context::ExecutorWorkerContext::execute_sync_task
             at ./src/query/service/src/pipelines/executor/executor_worker_context.rs:82:9
  23: databend_query::pipelines::executor::executor_worker_context::ExecutorWorkerContext::execute_task
             at ./src/query/service/src/pipelines/executor/executor_worker_context.rs:73:46
  24: databend_query::pipelines::executor::pipeline_executor::PipelineExecutor::execute_single_thread
             at ./src/query/service/src/pipelines/executor/pipeline_executor.rs:376:45
  25: databend_query::pipelines::executor::pipeline_executor::PipelineExecutor::execute_threads::{{closure}}::{{closure}}
             at ./src/query/service/src/pipelines/executor/pipeline_executor.rs:332:27
  26: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/core/src/panic/unwind_safe.rs:271:9
  27: std::panicking::try::do_call
             at /rustc/8b35c0bb0f833c0077dc57006eb317edde2a2d1e/library/std/src/panicking.rs:500:40
```

### How to Reproduce?

```sql
mysql> create table tt1(a boolean not null, b int not null, c int64 not null, d string not null);
Query OK, 0 rows affected (0.03 sec)

mysql> create table tt2(b int not null);
Query OK, 0 rows affected (0.03 sec)

mysql> insert into tt1 values
    -> (true,826203185,-3958367336395157675,'Jd695'),
    -> (true,-1439480114,-4881196055243782954,'Lzcda'),
    -> (false,733863457,6458421525263813535,'m8maS'),
    -> (true,-739644074,4390046195034875158,'zWvEj'),
    -> (true,-1611495410,-2002171580743726810,'rtOs8');
Query OK, 5 rows affected (0.03 sec)

mysql> insert into tt2 values
    -> (-1448403955),
    -> (1978682043),
    -> (775979323),
    -> (-2057711809),
    -> (183393966),
    -> (618200641);
Query OK, 6 rows affected (0.03 sec)

mysql> SELECT string_agg(regexp_substr(tt1.d, '-', tt1.c)), tt1.a FROM tt1 FULL OUTER JOIN tt2 USING(b) GROUP BY ROLLUP (tt1.a);
ERROR 1105 (HY000): PanicError. Code: 1104, Text = assertion failed: `(left == right)`
  left: `2`,
 right: `5`.
```

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by zenus**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

- Closes #12913

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12995)

### CODE REVIEW

**Comment by Dousir9** [comment]
Thanks for your contribution. I will review it tomorrow 😊.

File: src/query/expression/src/function.rs [hunk 1]
```diff
@@ -651,9 +653,18 @@ where F: Fn(&[ValueRef<AnyType>], &mut EvalContext) -> Value<AnyType> {
...
                 let result = match column {
                     Column::Nullable(box nullable_column) => {
-                        let validity = bitmap.into();
-                        let validity =
-                            common_arrow::arrow::bitmap::and(&nullable_column.validity, &validity);
+                        let mut before_map = bitmap.clone().into();
+                        let mut after_map = nullable_column.validity.clone();
+                        match bitmap.len().cmp(&after_map.len()) {
+                            Ordering::Greater => {
+                                after_map = expand_bitmap(&after_map, bitmap.len());
```
Comment (Dousir9) [inline src/query/expression/src/function.rs:660]: Maybe we can just use `bitmap.sliced(offset, length)` to get better performance.

Comment (zenus) [reply src/query/expression/src/function.rs:660]: yes, sliced is a good idea.

**Comment by sundy-li** [comment]
Hello, this is not the key place to fix the bug.  We must ensure the bitmaps have the same size in `and` and `or` ops.


The bug is that we forgot to push false into validity builder in `regexp_substr_fn`:
```
  if let Err(err) = regexp::validate_regexp_arguments("regexp_substr", pos, occur, None) {
            ctx.set_error(builder.len(), err);
            StringType::push_default(&mut builder);
            continue;
        }
```

**Comment by sundy-li** [comment]
There are other places with the same bugs in that file, you can fix them together

**Comment by zenus** [comment]
> StringType::push_default(&mut builder);

ok

**Review by sundy-li — APPROVED** [review]

**Review by Dousir9 — APPROVED** [review]

**Pull request merged**